src/analysis/convex/body.lean

```diff
@@ -0,0 +1,171 @@
+  .. convex_body.add_monoid }
+
+instance : has_smul ℝ≥0 (convex_body V) :=
+{ smul := λ c K, ⟨c • (K : set V), K.convex.smul _, K.is_compact.smul _, K.nonempty.smul_set⟩ }
+
+@[simp]
+lemma coe_smul (c : ℝ≥0) (K : convex_body V) : (↑(c • K) : set V) = c • (K : set V) := rfl
+
+instance : mul_action ℝ≥0 (convex_body V) :=
+{ to_has_smul := convex_body.has_smul,
+  one_smul := λ K, by { ext, simp only [coe_smul, one_smul] },
+  mul_smul := λ c d K, by { ext, simp only [coe_smul, mul_smul] } }
+
+instance : distrib_mul_action ℝ≥0 (convex_body V) :=
+{ to_mul_action := convex_body.mul_action,
+  smul_add := λ c K L, by { ext, simp only [coe_smul, coe_add, smul_add] },
+  smul_zero := λ c, by { ext, simp only [coe_smul, coe_zero, smul_zero] } }
+
+/--
+The convex bodies in a fixed space $V$ form a module over the nonnegative reals.
+-/
+instance : module ℝ≥0 (convex_body V) :=
+{ to_distrib_mul_action := convex_body.distrib_mul_action,
+  add_smul := λ c d K,
+  begin
+    ext1,
+    simp only [coe_smul, coe_add],
+    exact convex.add_smul K.convex (nnreal.coe_nonneg _) (nnreal.coe_nonneg _),
+  end,
+  zero_smul := λ K, by { ext1, exact set.zero_smul_set K.nonempty } }
+
+noncomputable instance : has_dist (convex_body V) :=
+{ dist := λ K L, metric.Hausdorff_dist (K : set V) L }
+
+lemma Hausdorff_edist_ne_top {K L : convex_body V} :
+emetric.Hausdorff_edist (K : set V) L ≠ ⊤ := by
+  apply_rules [metric.Hausdorff_edist_ne_top_of_nonempty_of_bounded, nonempty, bounded]
+
+/--
+Convex bodies in a fixed seminormed space $V$ form a pseudo-metric space.
+-/
+noncomputable instance : pseudo_metric_space (convex_body V) :=
+{ to_has_dist := convex_body.has_dist,
+  dist_self := λ _, metric.Hausdorff_dist_self_zero,
+  dist_comm := λ _ _, metric.Hausdorff_dist_comm,
+  dist_triangle := λ K L M, metric.Hausdorff_dist_triangle Hausdorff_edist_ne_top }
+
+end seminormed_space
+
+section normed_space
+
+variables {V : Type} [normed_add_comm_group V] [normed_space ℝ V]
+
+lemma is_closed {V : Type} [normed_add_comm_group V] [normed_space ℝ V] (K : convex_body V) :
+is_closed (K : set V) :=
+K.is_compact.is_closed
+
+lemma dist_eq_to_real_Hausdorff_edist (K L : convex_body V) :
+dist K L = (emetric.Hausdorff_edist (K : set V) L).to_real := rfl
+
+lemma of_real_dist_eq_Hausdorff_edist (K L : convex_body V) :
+ennreal.of_real (dist K L) = emetric.Hausdorff_edist (K : set V) L :=
+ennreal.of_real_to_real Hausdorff_edist_ne_top
+
+/--
+Convex bodies in a fixed normed space $V$ form a metric space.
+-/
+noncomputable instance : metric_space (convex_body V) :=
+{ to_pseudo_metric_space := convex_body.pseudo_metric_space,
+  eq_of_dist_eq_zero := λ K L hd,
+  begin
+    ext1,
+    rw [←emetric.Hausdorff_edist_zero_iff_eq_of_closed],
+    any_goals {apply is_closed},
+    rw [dist_eq_to_real_Hausdorff_edist] at hd,
+    replace hd := congr_arg ennreal.of_real hd,
+    simpa only [ennreal.of_real_zero, ennreal.of_real_to_real Hausdorff_edist_ne_top] using hd,
```

Comment:
Golf:
```suggestion
    exact (K.is_compact.is_closed.Hausdorff_dist_zero_iff_eq
      L.is_compact.is_closed Hausdorff_edist_ne_top).mp hd,
```
The above two lemmas are no longer used but I think they may be useful elsewhere.

Reply:
Very nice!